### PR TITLE
TDS8 Add Server Certificate Support

### PR DESCRIPTION
**Pull request by lcheunglci**
This adds the parameter `Server Certificate` to the connection string.  The server certificate is a local path to the SSL Certificate that has been download from the server.  This certificate will be manually validated during the SSL Handshake.  When HostNameInCertificate is specified with ServerCertificate, HNIC will take precedence.   I added very basic functional test to check for the connection string parameter.  Additional manual tests will be added in a later PR.

### CODE REVIEW

File: doc/snippets/Microsoft.Data.SqlClient/SqlConnection.xml [hunk 1]
```diff
@@ -542,6 +542,7 @@ End Module
...
 |Failover Partner SPN<br /><br /> -or-<br /><br /> FailoverPartnerSPN|N/A|The SPN for the failover partner. The default value is an empty string, which causes SqlClient to use the default, driver-generated SPN.<br /><br /> (Only available in v5.0+)|  
 |Host Name In Certificate<br /><br /> -or-<br /><br />HostNameInCertificate|N/A|The host name to use when validating the server certificate. When not specified, the server name from the Data Source is used for certificate validation.<br /><br /> (Only available in v5.0+)|
+|Server Certificate<br /><br /> -or-<br /><br />ServerCertificate|N/A|The path to a certificate file to match against the SQL Server TLS/SSL certificate. The accepted certificate formats are PEM, DER, and CER. If specified, the SQL Server certificate is checked by seeing if the ServerCertificate provided is an exact match.<br /><br /> (Only available in v5.0+)|
```
Comment (ErikEJ) [inline doc/snippets/Microsoft.Data.SqlClient/SqlConnection.xml:545]: by seeing => by verifying ?

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNITcpHandle.cs [hunk 2]
```diff
@@ -121,7 +122,8 @@ public override int ProtocolVersion
         /// <param name="cachedFQDN">Key for DNS Cache</param>
         /// <param name="pendingDNSInfo">Used for DNS Cache</param>
         /// <param name="tlsFirst">Support TDS8.0</param>
-        /// <param name="hostNameInCertificate">Host Name in Certoficate</param>
+        /// <param name="hostNameInCertificate">Host Name in Certificate</param>
+        /// <param name="serverCert">Used for the path to the Server Certificate</param>
```
Comment (ErikEJ) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNITcpHandle.cs:126]: serverCert => serverCertificatePath ?

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlConnectionString.cs [hunk 3]
```diff
@@ -91,8 +92,12 @@ internal static class KEY
             internal const string Context_Connection = DbConnectionStringKeywords.ContextConnection;
             internal const string Current_Language = DbConnectionStringKeywords.CurrentLanguage;
             internal const string Data_Source = DbConnectionStringKeywords.DataSource;
+
+            // Encrpytion related
```
Comment (ErikEJ) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlConnectionString.cs:96]: "Encrypt related"

File: src/Microsoft.Data.SqlClient/netcore/ref/Microsoft.Data.SqlClient.cs [hunk 4]
```diff
@@ -1033,6 +1033,11 @@ public sealed partial class SqlConnectionStringBuilder : System.Data.Common.DbCo
         [System.ComponentModel.DisplayNameAttribute("Host Name In Certificate")]
         [System.ComponentModel.RefreshPropertiesAttribute(System.ComponentModel.RefreshProperties.All)]
         public string HostNameInCertificate { get { throw null; } set { } }
+        /// <include file='../../../../doc/snippets/Microsoft.Data.SqlClient/SqlConnectionStringBuilder.xml' path='docs/members[@name="SqlConnectionStringBuilder"]/ServerCertificate/*'/>
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/netcore/ref/Microsoft.Data.SqlClient.cs:1036]: Please include changes in netfx > ref project as well. Also make sure you do add this property description in `doc/snippets/Microsoft.Data.SqlClient/SqlConnectionStringBuilder.xml`. 

I think it also got missed for HostNameInCertificate, and shows blank here: https://learn.microsoft.com/en-us/dotnet/api/microsoft.data.sqlclient.sqlconnectionstringbuilder?view=sqlclient-dotnet-standard-5.0
Please fix that too.

File: doc/snippets/Microsoft.Data.SqlClient/SqlConnectionStringBuilder.xml [hunk 5]
```diff
@@ -787,7 +806,7 @@ Connections are considered the same if they have the same connection string. Dif
 |Context Connection(Obsolete)|False|  
 |Current Language|Empty string|  
 |Data Source|Empty string|  
-|Encrypt|False|  
+|Encrypt|True|
```
Comment (David-Engel) [inline doc/snippets/Microsoft.Data.SqlClient/SqlConnectionStringBuilder.xml:809]:
```suggestion
|Encrypt|False in versions prior to 4.0, True in versions 4.0 and up|  
```

File: doc/snippets/Microsoft.Data.SqlClient/SqlConnectionStringBuilder.xml [hunk 6]
```diff
@@ -481,6 +481,25 @@ This property corresponds to the "FailoverPartnerSPN" and "Failover Partner SPN"
...
+        </value>
+        <remarks>
+          <format type="text/markdown">
+            <![CDATA[  
+
+## Remarks  
+This property corresponds to the "HostNameInCertificiate" and "Host Name in Certificiate" keys within the connection string. 
+
+> [!NOTE]
+> This property only applies when using `Encrypt` in <xref:Microsoft.Data.SqlClient.SqlConnectionEncryptOption.Strict%2A> mode, otherwise it is ignored.
```
Comment (David-Engel) [inline doc/snippets/Microsoft.Data.SqlClient/SqlConnectionStringBuilder.xml:497]: HNIC should apply in Mandatory or Strict mode. How does the implementation behave?

Comment (lcheunglci) [reply doc/snippets/Microsoft.Data.SqlClient/SqlConnectionStringBuilder.xml:497]: You're right.  I misinterpreted this feature and just verified that mandatory with the HNIC works and doesn't require the `TrustServerCertificate=true` to connect. Thanks!

Comment (lcheunglci) [reply doc/snippets/Microsoft.Data.SqlClient/SqlConnectionStringBuilder.xml]: Also verified the same for the `ServerCertificate`

File: doc/snippets/Microsoft.Data.SqlClient/SqlConnection.xml [hunk 7]
```diff
@@ -542,6 +542,7 @@ End Module
...
 |Failover Partner SPN<br /><br /> -or-<br /><br /> FailoverPartnerSPN|N/A|The SPN for the failover partner. The default value is an empty string, which causes SqlClient to use the default, driver-generated SPN.<br /><br /> (Only available in v5.0+)|  
 |Host Name In Certificate<br /><br /> -or-<br /><br />HostNameInCertificate|N/A|The host name to use when validating the server certificate. When not specified, the server name from the Data Source is used for certificate validation.<br /><br /> (Only available in v5.0+)|
+|Server Certificate<br /><br /> -or-<br /><br />ServerCertificate|N/A|The path to a certificate file to match against the SQL Server TLS/SSL certificate. The accepted certificate formats are PEM, DER, and CER. If specified, the SQL Server certificate is checked by verifying if the ServerCertificate provided is an exact match.<br /><br /> (Only available in v5.0+)|
```
Comment (David-Engel) [inline doc/snippets/Microsoft.Data.SqlClient/SqlConnection.xml:545]:
```suggestion
|Server Certificate<br /><br /> -or-<br /><br />ServerCertificate|N/A|The path to a certificate file to match against the SQL Server TLS/SSL certificate. The accepted certificate formats are PEM, DER, and CER. If specified, the SQL Server certificate is checked by verifying if the ServerCertificate provided is an exact match.<br /><br /> (Only available in v5.1+)|
```

File: src/Microsoft.Data.SqlClient/netfx/src/Resources/Strings.Designer.cs [hunk 8]
```diff
@@ -5559,6 +5559,15 @@ internal class Strings {
             }
         }
         
+        /// <summary>
+        ///   Looks up a localized string similar to When true, enables usage of the Asynchronous functionality in the .NET Framework Data Provider..
+        /// </summary>
+        internal static string DbConnectionString_AsynchronousProcessing {
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/netfx/src/Resources/Strings.Designer.cs:5565]: `Asynchronous Processing` has been dropped from netfx.

File: src/Microsoft.Data.SqlClient/netfx/src/Resources/Strings.Designer.cs [hunk 9]
```diff
@@ -8883,6 +8910,15 @@ internal class Strings {
             }
         }
         
+        /// <summary>
+        ///   Looks up a localized string similar to This command requires an asynchronous connection. Set &quot;Asynchronous Processing=true&quot; in the connection string..
+        /// </summary>
+        internal static string SQL_AsyncConnectionRequired {
+            get {
+                return ResourceManager.GetString("SQL_AsyncConnectionRequired", resourceCulture);
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/netfx/src/Resources/Strings.Designer.cs:8918]: same note here.

File: src/Microsoft.Data.SqlClient/netfx/src/Resources/Strings.Designer.cs [hunk 10]
```diff
@@ -10566,6 +10602,15 @@ internal class Strings {
             }
         }
         
+        /// <summary>
+        ///   Looks up a localized string similar to State of connection, synchronous or asynchronous.  &apos;Asynchronous Processing=x&apos; in the connection string..
+        /// </summary>
+        internal static string SqlConnection_Asynchronous {
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/netfx/src/Resources/Strings.Designer.cs:10608]: same note here.

**Comment by JRahnama** [comment]
the failure seems not relevant .Connection Open() without retries took longer than expected (2 seconds) on Mac machine.

**Comment by lcheunglci** [comment]
The mac failures seem to intermittently occur on the pipeline in all the PRs recently, so I just have to re-run it until it succeeds.

**Review by JRahnama — APPROVED** [review]

**Pull request merged**